tests/bin/difference.py

```diff
@@ -0,0 +1,406 @@
+
+        logging.info("Installing viewer %s into virtual environment %s", viewer_commit, venv_root)
+        run(["git", "checkout", viewer_commit], cwd=viewer_repo, env=env)
+        run(["make", "install"], cwd=viewer_repo, env=env)
+
+        venv_viewer = venv_bin / "cbmc-viewer"
+        return venv_viewer
+    except UserWarning as error:
+        logging.info("Failed to install viewer %s into virtual environment %s",
+                     viewer_commit, venv_root)
+        logging.info(error)
+        raise
+
+    finally:
+        logging.info("Restoring viewer repository to %s", branch)
+        run(["git", "checkout", branch], cwd=viewer_repo)
+        try:
+            run(["git", "stash", "pop"], cwd=viewer_repo)
+        except UserWarning:
+            # Probably means prior 'git stash' found nothing to stash
+            # Consider checking `git status --porcelain` to cover this case
+            pass
+
+################################################################
+# Reconstruct the litani add-job commands used to run viewer
+
+def load_run_json(proof_root):
+    """Load run.json generated by running run-cbmc-proofs.py in proof_root"""
+
+    run_json = run(['litani', 'dump-run'], cwd=proof_root)
+    return json.loads(' '.join(run_json))
+
+def viewer_wrappers(run_json):
+    """Litani wrappers for viewer jobs in run.json"""
+
+    for pipeline in run_json["pipelines"]:
+        for ci_stage in pipeline["ci_stages"]:
+            for job in ci_stage["jobs"]:
+                wrapper = job["wrapper_arguments"]
+                if wrapper["ci_stage"] == "report":
+                    yield wrapper
+
+def add_job_command(wrapper):
+    """Reconstruct a litani add-job command from its wrapper in run.json"""
+
+    return {
+        "--command": wrapper["command"],
+        "--inputs": wrapper["inputs"],
+        "--outputs": wrapper["outputs"],
+        "--pipeline-name": wrapper["pipeline_name"],
+        "--stdout-file": wrapper["stdout_file"],
+        "--interleave-stdout-stderr": wrapper["interleave_stdout_stderr"],
+        "--ci-stage": wrapper["ci_stage"],
+        "--description": wrapper["description"]
+    }
+
+def viewer_add_job_commands(proof_root):
+    """The litani add-job commands used to invoke viewer"""
+
+    run_json = load_run_json(proof_root)
+    wrappers = viewer_wrappers(run_json)
+    return [add_job_command(wrapper) for wrapper in wrappers]
+
+################################################################
+# Update a viewer job to use a different viewer and a different report directory
+
+def viewer_options_without_reportdir(viewer_command):
+    """Strip --reportdir from command line options used to invoke viewer."""
+
+    command = re.sub(r'\s+', ' ', viewer_command.strip())
+
+    # strip path used to invoke viewer from front of command
+    options = command.split(' ', 1)[-1]
+    # strip --reportdir option from within command
+    options = re.sub(r'--reportdir\s+[^\s]+', '', options)
```

Comment:
Nobody's mischievous enough to pass a report directory with spaces in the path name, right? 🙃 because then we really would need `shlex`. I think this is fine.

Reply:
Yeah, it would fail on Windows, but this is intended for regression testing on GitHub.